core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java

```diff
@@ -888,11 +865,27 @@ static boolean shouldIgnoreOrRejectNewClusterState(ESLogger logger, ClusterState
         }
     }
 
+    /**
+     * In the case we follow an elected master the new cluster state needs to have the same elected master
+     * This method checks for this and throws an exception if needed
+     */
+    public static void rejectNewClusterStateIfNeeded(ESLogger logger, DiscoveryNodes currentNodes, ClusterState newClusterState) {
+        if (currentNodes.masterNodeId() == null) {
+            return;
+        }
+        if (!currentNodes.masterNodeId().equals(newClusterState.nodes().masterNodeId())) {
+            logger.warn("received a cluster state from a different master then the current one, rejecting (received {}, current {})", newClusterState.nodes().masterNode(), currentNodes.masterNode());
```

Comment:
maybe instead log: `received a cluster state the follows a different elected master than we are following`?